spring-integration-ip/src/main/java/org/springframework/integration/ip/udp/DatagramPacketMessageMapper.java

```diff
@@ -142,7 +145,23 @@ public DatagramPacket fromMessage(Message<?> message) throws Exception {
 			buffer.put(bytes);
 			bytes = buffer.array();
 		}
-		return new DatagramPacket(bytes, bytes.length);
+		return getDatagramFromHeader(message, bytes, bytes.length);
+	}
+
+	private DatagramPacket getDatagramFromHeader(Message<?> message, byte[] bytes,
+												 int length) throws UnknownHostException {
+		InetAddress inetAddress = null;
+		if (message.getHeaders().get(IpHeaders.IP_ADDRESS, String.class) != null) {
+			inetAddress = InetAddress.getByName(message.getHeaders()
+					.get(IpHeaders.IP_ADDRESS, String.class));
```

Comment:
Sending packet has to have destination address. It can't contain receiver address. When we use socket-expression to pass reference from inbound adapter that means that messages from this inbound adapter will contain headers with client address = destination address.
When we use outbound adapter in a standard way (without socket-expression) w don't have these headers and thats why we have to populate outbound host/port parameters.
I'm not sure if I understand your question :/ It's late today and I have to take some sleep now. BTW - I'm travelling this weekend so I will be able to code from monday. But I'll read/write messages.